### PR TITLE
Add enumerations

### PR DESCRIPTION
**Pull request by anastasop**
The TileDB enumerations PR can be found at https://github.com/TileDB-Inc/TileDB/pull/4051

A usage example can be found at https://github.com/TileDB-Inc/TileDB/blob/dev/examples/cpp_api/enumerations.cc

### CODE REVIEW

File: enumeration.go [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+		return TILEDB_UINT64
+	case reflect.Bool:
+		return TILEDB_BOOL
+	default:
+		panic("can't get here")
+	}
+}
+
+// NewEnumeration creates an enumeration with name and ordered or not values.
+func NewEnumeration[T EnumerationType](tdbCtx *Context, name string, ordered bool, values []T) (*Enumeration, error) {
```
[the remark this answers is not in the document]
Comment (anastasop) [reply enumeration.go]: Chose 2 methods

File: enumeration.go [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+		for _, v := range values {
+			data = append(data, reflect.ValueOf(v).String()...)
+			offsets = append(offsets, currOffset)
+			currOffset += uint64(reflect.ValueOf(v).Len())
+		}
+		cCellNum = C.uint32_t(TILEDB_VAR_NUM)
+		cData = reflect.ValueOf(data).UnsafePointer()
+		cDataLen = C.uint64_t(dataSize)
+		cOffsets = reflect.ValueOf(offsets).UnsafePointer()
+		cOffsetsLen = C.uint64_t(len(values) * int(reflect.TypeOf(uint64(0)).Size()))
```
[the remark this answers is not in the document]
Comment (anastasop) [reply enumeration.go:110]: Good catch

File: enumeration.go [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+		return fmt.Errorf("Error path already %s exists", path)
+	}
+
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	cMode := C.CString("w")
+	defer C.free(unsafe.Pointer(cMode))
+
+	cFile := C.fopen(cPath, cMode)
```
[the remark this answers is not in the document]
Comment (anastasop) [reply enumeration.go:222]: Unfortunately the API requires a C.FILE not a file descriptor. You cannot get this from os package

File: enumeration.go [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+	var cDataSize C.uint64_t
+	ret := C.tiledb_enumeration_get_data(e.context.tiledbContext, e.tiledbEnum, &cData, &cDataSize)
+	if ret != C.TILEDB_OK {
+		return nil, fmt.Errorf("Error getting data: %s", e.context.LastError())
+	}
+
+	if typ != TILEDB_STRING_ASCII {
+		switch typ {
+		case TILEDB_BOOL:
+			return unsafe.Slice((*bool)(cData), cDataSize), nil
```
[the remark this answers is not in the document]
Comment (anastasop) [reply enumeration.go]: Make copies for safety

**Review by thetorpedodog — APPROVED** [review]
Of course awaiting the release to make sure it works for real, but otherwise this all looks good!

**Comment by anastasop** [comment]
@thetorpedodog Setup github actions for 2.17.0-rc0 and works. Merge is pending until we test with a 2.17 release

**Comment by anastasop** [comment]
@thetorpedodog Rebased to use TileDB-2.17.0 release

**Comment by thetorpedodog** [comment]
still looks good

**Pull request merged**